java/arcs/core/analysis/ClaimDeduction.kt

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.analysis
+
+import arcs.core.data.expression.Expression
+
+internal typealias Path = List<String>
+internal typealias Paths = List<Path>
+internal typealias ClaimDerivations = Map<Path, Set<Path>>
+
+fun <E> MutableList<E>.push(element: E) = add(element)
+fun <E> MutableList<E>.pop(): E = removeAt(size - 1)
+fun <E> MutableList<E>.peek(): E = this[count()]
+fun <E> MutableList<E>.popOrDefault(default: E): E = if (size == 0) default else pop()
+fun <E> MutableList<E>.peekOrDefault(default: E): E = if (size == 0) default else peek()
+
+/** A visitor that accumulates [ClaimDerivations] from a Paxel [Expression]. */
+class ClaimDeducer : Expression.Visitor<ClaimDerivations> {
+
+    /** Stack of intermediate [Path]s; they have no derivation (yet). */
+    /* internal */ val stack = mutableListOf<Paths>()
+
+    override fun <E, T> visit(expr: Expression.UnaryExpression<E, T>): ClaimDerivations {
+        TODO("Not yet implemented")
+    }
+
+    override fun <L, R, T> visit(expr: Expression.BinaryExpression<L, R, T>): ClaimDerivations {
+        val result = expr.left.accept(this) + expr.right.accept(this)
+
+        stack.push(stack.pop() + stack.pop())
+
+        return result
+    }
+
+    override fun <T> visit(expr: Expression.FieldExpression<T>): ClaimDerivations {
+        // Base case: unqualified field must refer to handle connection.
+        if (expr.qualifier == null) {
+            stack.push(listOf(listOf(expr.field)))
```

Comment:
Instead of keeping a stack, you might want to create a separate visitor to collect field paths.